### ISSUE TITLE
ERROR when upgrade storm topology with KafkaSpout

### ISSUE THREAD

**Issue by mycFelix** [twitter/heron#1146]
Hi,

I tried to upgrade Storm topology with KafkaSpout following [http://twitter.github.io/heron/docs/upgrade-storm-to-heron/](http://twitter.github.io/heron/docs/upgrade-storm-to-heron/) guide.

The code is very simple and runs well in Storm, which looks like:

````Java
package heron.kafka;

import backtype.storm.LocalCluster;
import backtype.storm.StormSubmitter;
import backtype.storm.generated.AlreadyAliveException;
import backtype.storm.topology.TopologyBuilder;
import com.twitter.heron.api.Config;
import storm.kafka.BrokerHosts;
import storm.kafka.KafkaSpout;
import storm.kafka.SpoutConfig;
import storm.kafka.ZkHosts;

import java.util.Arrays;

/**
 * Created by Felix on 16/7/25.
 */
public class KafkaBaseTest {
    public static void main(String[] args) throws AlreadyAliveException, backtype.storm.generated.InvalidTopologyException {

        String zkConnString = "127.0.0.1:2181";
        String topicName = "test";
        String zkRoot = "/felix/brokers";

        BrokerHosts hosts = new ZkHosts(zkConnString, zkRoot);
        SpoutConfig spoutConfig = new SpoutConfig(hosts, topicName, "/kafka/" + topicName, "felix");
        spoutConfig.scheme = new PrintScheme();

        spoutConfig.zkServers = Arrays.asList("127.0.0.1");
        spoutConfig.zkPort = 2181;

        KafkaSpout kafkaSpout = new KafkaSpout(spoutConfig);

        TopologyBuilder builder = new TopologyBuilder();
        builder.setSpout("kafka", kafkaSpout);

        Config conf = new Config();
        // must set,or it will cause null exceptions
        conf.put("storm.zookeeper.session.timeout", 20000);
        conf.put("storm.zookeeper.connection.timeout", 15000);
        conf.put("storm.zookeeper.retry.times", 5);
        conf.put("storm.zookeeper.retry.interval", 1000);

        StormSubmitter.submitTopology("Kafka-Base-Test", conf, builder.createTopology());

    }
}
````

There are two problems.
*  I got an `NoClassDefFoundError: storm/trident/spout/ISpoutPartition` exception. When I put `storm-core` jar to my jar, the error disappeared, and the topology was running well. Logs are the following:
````
[2016-07-25 17:26:34 +0800] com.twitter.heron.instance.HeronInstance 严重:  Exception caught in thread: SlaveThread with id: 13 
java.lang.NoClassDefFoundError: storm/trident/spout/ISpoutPartition
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at storm.kafka.trident.GlobalPartitionInformation.getOrderedPartitions(GlobalPartitionInformation.java:54)
	at storm.kafka.KafkaUtils.calculatePartitionsForTask(KafkaUtils.java:214)
	at storm.kafka.ZkCoordinator.refresh(ZkCoordinator.java:80)
	at storm.kafka.ZkCoordinator.getMyManagedPartitions(ZkCoordinator.java:69)
	at storm.kafka.KafkaSpout.nextTuple(KafkaSpout.java:135)
	at backtype.storm.topology.IRichSpoutDelegate.nextTuple(IRichSpoutDelegate.java:63)
	at com.twitter.heron.instance.spout.SpoutInstance.produceTuple(SpoutInstance.java:271)
	at com.twitter.heron.instance.spout.SpoutInstance.access$100(SpoutInstance.java:42)
	at com.twitter.heron.instance.spout.SpoutInstance$1.run(SpoutInstance.java:176)
	at com.twitter.heron.common.basics.WakeableLooper.executeTasksOnWakeup(WakeableLooper.java:142)
	at com.twitter.heron.common.basics.WakeableLooper.runOnce(WakeableLooper.java:74)
	at com.twitter.heron.common.basics.WakeableLooper.loop(WakeableLooper.java:64)
	at com.twitter.heron.instance.Slave.run(Slave.java:169)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: storm.trident.spout.ISpoutPartition
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 28 more
````

* The configurations like `storm.zookeeper.session.timeout`,`storm.zookeeper.connection.timeout`... etc must be set,or it will cause `null` exceptions when create `CuratorFramework` object in `ZkState` and `DynamicPartitionConnections` class. Logs are following:
````
[2016-07-25 15:26:39 +0800] com.twitter.heron.instance.HeronInstance 严重:  Exception caught in thread: SlaveThread with id: 13
        java.lang.NullPointerException
        at storm.kafka.KafkaSpout.open(KafkaSpout.java:80)
        at backtype.storm.topology.IRichSpoutDelegate.open(IRichSpoutDelegate.java:43)
        at com.twitter.heron.instance.spout.SpoutInstance.start(SpoutInstance.java:125)
        at com.twitter.heron.instance.Slave.startInstance(Slave.java:173)
        at com.twitter.heron.instance.Slave.handleNewAssignment(Slave.java:159)
        at com.twitter.heron.instance.Slave.access$200(Slave.java:40)
        at com.twitter.heron.instance.Slave$1.run(Slave.java:85)
        at com.twitter.heron.common.basics.WakeableLooper.executeTasksOnWakeup(WakeableLooper.java:142)
        at com.twitter.heron.common.basics.WakeableLooper.runOnce(WakeableLooper.java:74)
        at com.twitter.heron.common.basics.WakeableLooper.loop(WakeableLooper.java:64)
        at com.twitter.heron.instance.Slave.run(Slave.java:169)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
````

**Comment by kramasamy** [comment]
@mycFelix - that could be because of trident? currently heron does not support trident.

**Comment by mycFelix** [comment]
@kramasamy - hi~ I'm using `storm core api`,not `trident api`. 
* About the first question,`NoClassDefFoundError `. Maybe the logs show more details. Please notice this part:
````
   at storm.kafka.trident.GlobalPartitionInformation.getOrderedPartitions(GlobalPartitionInformation.java:54)
    at storm.kafka.KafkaUtils.calculatePartitionsForTask(KafkaUtils.java:214)
    at storm.kafka.ZkCoordinator.refresh(ZkCoordinator.java:80)
    at storm.kafka.ZkCoordinator.getMyManagedPartitions(ZkCoordinator.java:69)
    at storm.kafka.KafkaSpout.nextTuple(KafkaSpout.java:135)
````
The `storm.kafka.KafkaUtils` class invoked `calculatePartitionsForTask` method. And
the `GlobalPartitionInformation` class belongs to the package `storm.kafka.trident`,which aims to find  kafka partition informations.The `ISpoutPartition` interface,belongs to the package `storm.trident.spout`, was invoked by `GlobalPartitionInformation.getOrderedPartitions` method.
````
public class GlobalPartitionInformation implements Iterable<Partition>, Serializable {
……………………
public List<Partition> getOrderedPartitions() {
        List<Partition> partitions = new LinkedList<Partition>();
        for (Map.Entry<Integer, Broker> partition : partitionMap.entrySet()) {
            partitions.add(new Partition(partition.getValue(), partition.getKey()));
        }
        return partitions;
    }
````
````
package storm.kafka;

import com.google.common.base.Objects;
import storm.trident.spout.ISpoutPartition;

public class Partition implements ISpoutPartition {
……………………………………
````

The `storm-kafka-0.10.0` jar is included in my `assembly.xml`,`storm-core` isn't. 
````
<include>org.apache.storm:storm-kafka</include>
```` 


* About the second question. it will cause `null exceptions` when creates CuratorFramework object in ZkState and DynamicPartitionConnections class if the configurations about the `zookeeper` were not set. The following source code：
````
package storm.kafka;
public class ZkState {
    public static final Logger LOG = LoggerFactory.getLogger(ZkState.class);
    CuratorFramework _curator;

    private CuratorFramework newCurator(Map stateConf) throws Exception {
        Integer port = (Integer) stateConf.get(Config.TRANSACTIONAL_ZOOKEEPER_PORT);
        String serverPorts = "";
        for (String server : (List<String>) stateConf.get(Config.TRANSACTIONAL_ZOOKEEPER_SERVERS)) {
            serverPorts = serverPorts + server + ":" + port + ",";
        }

        /*
         must set,or it will cause null exceptions in heron
         */
        return CuratorFrameworkFactory.newClient(serverPorts,
                Utils.getInt(stateConf.get(Config.STORM_ZOOKEEPER_SESSION_TIMEOUT)),
                15000,
                new RetryNTimes(Utils.getInt(stateConf.get(Config.STORM_ZOOKEEPER_RETRY_TIMES)),
                        Utils.getInt(stateConf.get(Config.STORM_ZOOKEEPER_RETRY_INTERVAL))));
    }
````

````
package storm.kafka;
public class DynamicBrokersReader {
    public static final Logger LOG = LoggerFactory.getLogger(DynamicBrokersReader.class);
    private CuratorFramework _curator;
    private String _zkPath;
    private String _topic;

    public DynamicBrokersReader(Map conf, String zkStr, String zkPath, String topic) {
        Preconditions.checkNotNull(conf, "conf cannot be null");
        /*
         must set,or it will cause null exceptions in heron
         */
        this.validateConfig(conf);
        Preconditions.checkNotNull(zkStr, "zkString cannot be null");
        Preconditions.checkNotNull(zkPath, "zkPath cannot be null");
        Preconditions.checkNotNull(topic, "topic cannot be null");
        this._zkPath = zkPath;
        this._topic = topic;
…………………………………………………………………………

private void validateConfig(Map conf) {
        Preconditions.checkNotNull(conf.get("storm.zookeeper.session.timeout"), "%s cannot be null", new Object[]{"storm.zookeeper.session.timeout"});
        Preconditions.checkNotNull(conf.get("storm.zookeeper.connection.timeout"), "%s cannot be null", new Object[]{"storm.zookeeper.connection.timeout"});
        Preconditions.checkNotNull(conf.get("storm.zookeeper.retry.times"), "%s cannot be null", new Object[]{"storm.zookeeper.retry.times"});
        Preconditions.checkNotNull(conf.get("storm.zookeeper.retry.interval"), "%s cannot be null", new Object[]{"storm.zookeeper.retry.interval"});
    }

````

**Comment by maosongfu** [comment]
@mycFelix Thanks to tell us this; it is super helper. 
So what you think next steps are? Since it is more an issue of storm kafka-spout (it relies on some config set in-explicitly), and we could barely do a lot of things on that. One option can be have our own Kafka Spout (can be modified from existing storm one) and put into heron-contrib repo or folder.

**Comment by mycFelix** [comment]
@maosongfu - Thank you for taking your time.

Like what you said,`it is more an issue of storm kafka-spout`. Maybe I'm not the only person who met those problems.
I think it should be considered carefully before any decision is made,cause one of the best advantages of heron for most developers is upgrading from storm `unpainful`.

If there has something the community has to do, `heron-kafka-spout` probably is the best option. 

However,it is my personal view.There might be a lot of `kafka-spout` users and their voice should be considered too.

**Comment by kramasamy** [comment]
@mycFelix - if you could contribute for this, it will be awesome.

**Comment by mycFelix** [comment]
@kramasamy - Thank you. 
I would like to but I'm afraid that i couldn't handle it.
I'm very happy to be the first one to test this function.

**Comment by kramasamy** [comment]
OK. @caofangkun - can you help us on this?

**Comment by ZacharyThomas** [comment]
So is it safe to say that the storm kafka spout isn't backwards compatible currently? Is there any alternative?

**Comment by kramasamy** [comment]
Actually we have internal Kafka Spout which works cleanly for us. The storm kafka spout has diverged. We have a pull request that we need to work on 

https://github.com/twitter/heron/pull/804

to get the Kafka spout in. We are bumping this a priority for the next release - 0.14.3.

**Comment by objmagic** [comment]
@mycFelix I am working on this. I will greatly appreciate if you can provide a repository/zip which I can use to reproduce the problem.

**Comment by objmagic** [comment]
@mycFelix can I have your `pom.xml`?

**Comment by mycFelix** [comment]
@objmagic - Yes. I would like to. It is early in the morning in my country, around 04:00am. The pom is in my company's computer. I will upload it as soon as I get work.

**Comment by objmagic** [comment]
@mycFelix No problem. Can you also provide me your e-mail address so I can invite you to our Slack dev group?

**Comment by mycFelix** [comment]
@mycFelix - Sure. Thank you. It's my honor. My e-mail:myc3366@gmail.com

**Comment by mycFelix** [comment]
@objmagic - Hi, sorry for your waiting.

There would be more details might help you to reproduce problem on [heron-study](https://github.com/mycFelix/heron-study), and the `pom.xml` could be found on [pom](https://github.com/mycFelix/heron-study/blob/master/pom.xml). Please let me know if there is anything I could help.

Thank you for your invitation.It's my honor. My e-mail address: mys3366#gmail.com

**Comment by nlu90** [comment]
Hi @mycFelix, can you try to use the heron-api and heron-storm jars in the following zip file?

[heron-api-storm.zip](https://github.com/twitter/heron/files/421066/heron-api-storm.zip)

And let me know if there's any problem

**Comment by mycFelix** [comment]
@nlu90 -- Hi，It means heron-kafka-spout has been patched in `heron-api.jar` or `heron-storm.jar`?

**Comment by kramasamy** [comment]
@mycFelix - @nlu90 has patched the heron api to fix the missing API so that storm-kafka spout can work. Please try it out and let us know. For the bigger picture, we are discussing whether to align the spouts to storm distribution or create specific heron spouts.

**Comment by mycFelix** [comment]
@kramasamy @nlu90 -- Thank you guys.
I run test code. It seems we need one more function `Utils.getBoolean()` ,the error logs are following:

````
java.lang.NoSuchMethodError: org.apache.storm.utils.Utils.getBoolean(Ljava/lang/Object;Z)Z
	at org.apache.storm.kafka.DynamicBrokersReader.<init>(DynamicBrokersReader.java:59)
	at org.apache.storm.kafka.trident.ZkBrokerReader.<init>(ZkBrokerReader.java:43)
	at org.apache.storm.kafka.KafkaUtils.makeBrokerReader(KafkaUtils.java:58)
	at org.apache.storm.kafka.KafkaSpout.open(KafkaSpout.java:77)
	at org.apache.storm.topology.IRichSpoutDelegate.open(IRichSpoutDelegate.java:47)
	at com.twitter.heron.simulator.instance.SpoutInstance.start(SpoutInstance.java:124)
	at com.twitter.heron.simulator.executors.InstanceExecutor.startInstance(InstanceExecutor.java:212)
	at com.twitter.heron.simulator.executors.InstanceExecutor.run(InstanceExecutor.java:181)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
````
And the `storm-zookeeper` configurations must be set.

**Comment by kramasamy** [comment]
@nlu90 - can you please add that method and give a custom jar for @mycFelix to try it out?

**Comment by mycFelix** [comment]
@nlu90 - It's my pleasure to do any test on it. if you want, we can contact each other using email address, in case there has more problems. My email address: myc3366#gmail.com

**Comment by kramasamy** [comment]
@mycFelix - I have invited you to our heron slack group. Please check your gmail for an invitation - it might help speed up things.

**Comment by nlu90** [comment]
@mycFelix Thanks for the feedback! Here is the new customized jar:
[heron-api-storm.zip](https://github.com/twitter/heron/files/423409/heron-api-storm.zip)

**Comment by mycFelix** [comment]
@kramasamy - Thank you for your invitation

@nlu90 - I run the test code using new jars. It works when I try to connect kafka,but fails to read data. My code is following:

````
BrokerHosts hosts = new ZkHosts(zkConnString, zkRoot);
SpoutConfig spoutConfig = new SpoutConfig(hosts, topicName, "/kafka/" + topicName, "felix");
spoutConfig.scheme = new PrintScheme();

……………………

import org.apache.storm.spout.MultiScheme;
import org.apache.storm.tuple.Fields;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class PrintScheme implements MultiScheme {
    @Override
    public Iterable<List<Object>> deserialize(byte[] bytes) {
        String message = new String(bytes, StandardCharsets.UTF_8);
        System.out.println("kafka=======>" + message);
        List<List<Object>> result = new ArrayList<>();
        
````

The error is following:
````
java.lang.NoSuchMethodError: org.apache.storm.spout.MultiScheme.deserialize(Ljava/nio/ByteBuffer;)Ljava/lang/Iterable;
	at org.apache.storm.kafka.KafkaUtils.generateTuples(KafkaUtils.java:235)
	at org.apache.storm.kafka.PartitionManager.next(PartitionManager.java:161)
	at org.apache.storm.kafka.KafkaSpout.nextTuple(KafkaSpout.java:135)
	at org.apache.storm.topology.IRichSpoutDelegate.nextTuple(IRichSpoutDelegate.java:67)
	at com.twitter.heron.simulator.instance.SpoutInstance.produceTuple(SpoutInstance.java:263)
````

And the storm-kafka `KafkaConfig.scheme` seems need `org.apache.storm.spout.MultiScheme`. The error seems due to the interface difference.

**Comment by nlu90** [comment]
@mycFelix Yes, it's due to the different interfaces.

Tn storm, the interface of MultiScheme is `public Iterable<List<Object>> deserialize(ByteBuffer ser);`,
while in heron, we use `Iterable<List<Object>> deserialize(byte[] ser);` 

The two methods accept different type of parameters. There are 2 ways we can solve the probelm:
1. update the interface in Heron to use `ByteBuffer`; This may cause a lot of effort to migrate.
2. update the storm-kafka code to provide `byte[]` to the method; This requires we diverge from the storm kafka spout.

Let me know your thoughts about the problem.

**Comment by mycFelix** [comment]
@nlu90 -

**Comment by lukess** [comment]
What will be the possible solution of this issue?
a hack like https://github.com/elodina/heron/tree/kafka_spout or a new heron-kafka?

**Comment by mycFelix** [comment]
@lukess - Option 2,new heron-kafka  will be the possible solution of this issue. @nlu90 is working on it. And heron-kafka-spout will be coming soon.

**Comment by lukess** [comment]
@mycFelix thanks!

**Comment by nlu90** [comment]
#1299

**Comment by mycFelix** [comment]
@khushboo13 - hi, those are test jars. @nlu90 has new PR(#1299 & #1317) for heron-kafka-spout that support for storm-core-api. And we will have new jars to make it works soon.

**Comment by kramasamy** [comment]
this is fixed in #1299

**Comment by Yitian-Zhang** [comment]
I encountered the same question: [#3103](https://github.com/apache/incubator-heron/issues/3103).
